EIPS/transient-pricing.md

```diff
@@ -0,0 +1,55 @@
+---
+title: Decrease TLOAD/TSTORE pricing for common cases
+description: Improve the efficiency of TLOAD/TSTORE by introducing a quadratic(?) pricing model.
+author: @charles-cooper, @prestwich, @brockelmore
+discussions-to: none yet
+status: Draft
+type: Standards Track
+category: Core
+created: tbd
+requires: 1153
+---
+
+## Abstract
+
+Increase the efficiency of TLOAD/TSTORE for common use cases, while providing a pricing model to prevent DoS vectors.
+
+## Motivation
+
+EIP-1153 introduces a new storage region, termed "transient storage", which behaves like storage in the sense that it is word-addressed and persists between call frames, but unlike storage in the sense that it is wiped at the end of each transaction. During development of the 1153 specification, it was decided to match the pricing to be the same as warm storage loads and stores. This was for two reasons: conceptual simplicity of the EIP, and it also addressed concerns about two related DoS vectors: being able to allocate too much transient storage, and the cost of rolling back state in the case of reverts.
+
+One of the most important use cases that EIP-1153 enables is cheap reentrancy protection. In fact, if transient storage is cheap enough for the first few slots, reentrancy protection can be enabled by default at the language level without too much burden to users, while simultaneously preventing the largest - and most expensive! - class of smart contract vulnerabilities.
+
+Furthermore, it seems that transient storage is fundamentally overpriced. Its pricing does not interact with refunds, it only requires a new allocation on contract load (as opposed to memory, which requires a fresh allocation on every call), and has no interaction with the physical database.
+
+This EIP proposes a quadratic pricing model, which is cheaper for common cases (fewer than 33 slots are written per contract), while making DoS using transient storage prohibitively expensive.
+
+
+## Specification
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+The gas cost for TLOAD is proposed to be 9 gas. The gas cost for TSTORE is proposed to be 15 gas + `expansion_cost`, where expansion cost is calculated as 3 gas * the number of transient slots already allocated for this contract.
```

Comment:
are there any other opcodes that result in a lookup in a map? because the lookup is constant time but has some fixed overhead, and that would inform what the base gas cost should be

Reply:
also any color to add on how these constants were chosen would be helpful

Reply:
> are there any other opcodes that result in a lookup in a map? because the lookup is constant time but has some fixed overhead, and that would inform what the base gas cost should be

i think the only comparable opcode here would be warm storage access. note that sha3 is 36 gas to hash a single word, so that can be considered the upper bound on a hash function.

Reply:
see https://ethereum-magicians.org/t/eip-8158-reduce-transient-storage-pricing/18435/2 for benchmarks -- essentially, tload and tstore seem to perform like `G_low` and `G_mid`, respectively.